torch._assert: make it torch.jit.script'able

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47399 torch._assert: make it torch.jit.script'able**
* #47763 rename torch.Assert to torch._assert

Summary:

Currently `torch._assert` is not scriptable, which makes it not very useful for production code.  Adding a JIT override to address this.

Test Plan:

```
python test/test_utils.py TestAssert.test_assert_scriptable
python test/test_utils.py TestAssert.test_assert_true
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

vkuzo committed:
torch.Assert: move import out of function

Summary:

Looks like the current code is causing some `torch.jit.script` issues.
Moving the import out should fix it.

Test Plan: CI

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
make imports better on "torch.Assert: move import out of function"


Summary:

Looks like the current code is causing some `torch.jit.script` issues with models that use `torch.Assert`.
Moving the import out should help.

Test Plan: 

```
import torch                            
import torch.nn as nn                   
import torch.fx                         
                                        
class M(nn.Module):                     
    def __init__(self):                 
        super().__init__()              
                                        
    def forward(self, x):               
        torch.Assert(x.sum() > 0, 'foo')
        return x                        
                                        
m = M()                                 
gm = torch.fx.symbolic_trace(m)  
# before PR, failed with https://gist.github.com/vkuzo/a1ded845dd026865010bfa5a76ab67ad
# after PR, progresses further and fails with https://gist.github.com/vkuzo/27b09ceb5f6f3c17b13183df1abc5038       
gms = torch.jit.script(gm)              
```

We'll have to follow up with a separate PR to make it 100% scriptable.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

albanD reviewed (approved):
lgtm

vkuzo committed:
move to c++, part 1 on "torch.Assert: remove import from function"


Summary:

Looks like the current code is causing some `torch.jit.script` issues with models that use `torch.Assert`.
Moving the import out should help.

Test Plan: 

```
import torch                            
import torch.nn as nn                   
import torch.fx                         
                                        
class M(nn.Module):                     
    def __init__(self):                 
        super().__init__()              
                                        
    def forward(self, x):               
        torch.Assert(x.sum() > 0, 'foo')
        return x                        
                                        
m = M()                                 
gm = torch.fx.symbolic_trace(m)  
# before PR, failed with https://gist.github.com/vkuzo/a1ded845dd026865010bfa5a76ab67ad
# after PR, progresses further and fails with https://gist.github.com/vkuzo/27b09ceb5f6f3c17b13183df1abc5038       
gms = torch.jit.script(gm)              
```

We'll have to follow up with a separate PR to make it 100% scriptable.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] torch.Assert: attempt to make scriptable"


Summary:

Currently `torch.Assert` is not scriptable, which makes it not very useful for production code.  According to @jamesr66a , moving this to c++ op land will help with scriptability.  It's not 100% there yet, needs some review before further progress.

Current status:
```
# asserting things still works

(pytorch) [vasiliy@devgpu108.ash6 ~/local/pytorch] python test/test_utils.py TestAssert
Couldn't download test skip set, leaving all tests enabled...
.
----------------------------------------------------------------------
Ran 1 test in 0.178s

OK

# the overrides no longer work, TODO figure out why

(pytorch) [vasiliy@devgpu108.ash6 ~/local/pytorch] python test/test_fx.py TestFX.test_symbolic_trace_assert
Couldn't download test skip set, leaving all tests enabled...
E
======================================================================
ERROR: test_symbolic_trace_assert (__main__.TestFX)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_fx.py", line 734, in test_symbolic_trace_assert
    traced = symbolic_trace(m)
  File "/data/users/vasiliy/pytorch/torch/fx/symbolic_trace.py", line 215, in symbolic_trace
    return GraphModule(root if isinstance(root, torch.nn.Module) else torch.nn.Module(), Tracer().trace(root
))
  File "/data/users/vasiliy/pytorch/torch/fx/symbolic_trace.py", line 199, in trace
    self.create_node('output', 'output', (self.create_arg(fn(*args)),), {},
  File "test/test_fx.py", line 729, in forward
    torch.Assert(x.shape[1] > 4, message)
TypeError: Assert(): argument 'condition' (position 1) must be bool, not Proxy

```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

vkuzo commented:
@albanD , thanks for the initial review! The scope of this PR has changed, so adding a new set of reviewers.

vkuzo committed:
rebase on "torch.Assert: make it torch.jit.script'able"


Summary:

Currently `torch.Assert` is not scriptable, which makes it not very useful for production code.  According to @jamesr66a , moving this to c++ op land will help with scriptability.  This PR implements the change.

Note: with the current code the Assert is scriptable but the Assert is a no-op after being scripted.  Would love suggestions on how to address that (can be in future PR).

Test Plan:

```
python test/test_utils.py TestAssert.test_assert_scriptable
python test/test_utils.py TestAssert.test_assert_true
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -328,6 +328,10 @@
 
+- func: Assert(bool condition, str message) -> ()
+
+- func: Assert.Tensor(Tensor condition, str message) -> ()
```
Note: this would be the very first function in `native_functions.yaml` to start with a capital letter.

ezyang replied:
You know what I'm going paint this bikeshed. I don't want this function to be named Assert with a capital A. Python naming convention is that all functions start with lower case and classes are upper case, and this violates that. If I had to give a suggestion, maybe try `_assert` (not `assert_` as that would imply inplace, also bad.)

vkuzo replied:
sure, moved that to https://github.com/pytorch/pytorch/pull/47763

ezyang commented on aten/src/ATen/native/Assert.cpp:
```diff
@@ -0,0 +1,23 @@
+}
+
+void Assert(const Tensor& condition, std::string message) {
+  if (!at::native::all(condition).item<bool>()) {
```
Does this match the boolean coercion rules for tensor? I don't think so.

```
>>> import torch
>>> bool(torch.ones(2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Boolean value of Tensor with more than one value is ambiguous
```

vkuzo replied:
thanks, fixed

ezyang reviewed (changes requested):
Tensor version should match behavior of boolean conversion on Tensors.

vkuzo committed:
Update on "torch.Assert: make it torch.jit.script'able"


Summary:

Currently `torch.Assert` is not scriptable, which makes it not very useful for production code.  According to @jamesr66a , moving this to c++ op land will help with scriptability.  This PR implements the change.

Note: with the current code the Assert is scriptable but the Assert is a no-op after being scripted.  Would love suggestions on how to address that (can be in future PR).

Test Plan:

```
python test/test_utils.py TestAssert.test_assert_scriptable
python test/test_utils.py TestAssert.test_assert_true
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

vkuzo committed:
rebase on "torch._assert: make it torch.jit.script'able"


Summary:

Currently `torch._assert` is not scriptable, which makes it not very useful for production code.  According to @jamesr66a , moving this to c++ op land will help with scriptability.  This PR implements the change.

Note: with the current code the `_assert` is scriptable but is a no-op after being scripted.  Would love suggestions on how to address that (can be in future PR).

Test Plan:

```
python test/test_utils.py TestAssert.test_assert_scriptable
python test/test_utils.py TestAssert.test_assert_true
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/Assert.cpp:
```diff
@@ -0,0 +1,25 @@
+  }
+}
+
+void _assert(const Tensor& condition, std::string message) {
```
what is the behavior when `condition` is not a Scalar? or do we require the input to be a Scalar? can we add assert/doc for this?

vkuzo replied:
the `condition.is_nonzero()` function used later on in this function verifies that the tensor has a single element and throws an error otherwise.  So, for now we are piggybacking on those checks.  Lmk if we need something stronger.

jerryzh168 replied:
sounds good, please add a comment thanks!

eellison commented:
I think this probably shouldn't be registered in native_functions.yaml, and should be registered in `builtin_functions.cpp`. Ideally we wouldnt add new operators just to check a condition and throw an error. The operators as registered will also always get DCE'd and so not do anything. Registering it in builtin_functions.cpp may or may not work right now bc the builtin_functions get inlining into their calling graph before conversion to SSA happens, so there can be wonky interactions if the buitlin_functions have control flow. I don't think that applies in this case though because the if node shouldn't have any outputs. I'm going to write up an issue describing the control flow interaction..

vkuzo commented:
> I think this probably shouldn't be registered in native_functions.yaml, and should be registered in `builtin_functions.cpp`. Ideally we wouldnt add new operators just to check a condition and throw an error. The operators as registered will also always get DCE'd and so not do anything. Registering it in builtin_functions.cpp may or may not work right now bc the builtin_functions get inlining into their calling graph before conversion to SSA happens, so there can be wonky interactions if the buitlin_functions have control flow. I don't think that applies in this case though because the if node shouldn't have any outputs. I'm going to write up an issue describing the control flow interaction..

would functions registered there still get the `__torch_function__` override behavior?

eellison commented:
@vkuzo no, it's just for defining how something compiles. You would keep the existing python snippet you've defined in `torch/__init__.py` and I think everything would work.

vkuzo commented:
> @vkuzo no, it's just for defining how something compiles. You would keep the existing python snippet you've defined in `torch/__init__.py` and I think everything would work.

to clarify, do you mean this snippet?

```
    if type(condition) is not torch.Tensor and has_torch_function((condition,)):	
        return handle_torch_function(_assert, (condition,), condition, message)	
```

One issue is that the snippet above is not scriptable (hence this PR).  Or, lmk if you mean something different.

eellison commented:
Basically, this PR would be replaced by registering `torch._assert` as a builtin in `_builtins.py`:
`(torch.assert, "aten::_assert")`, 
and then registering in `builtin_functions.cpp`
```
def _assert(bool condition, str message):
    assert condition, message
```
(also add the tensor overload)

vkuzo commented:
> Basically, this PR would be replaced by registering `torch._assert` as a builtin in `_builtins.py`:
> `(torch.assert, "aten::_assert")`,
> and then registering in `builtin_functions.cpp`
> 
> ```
> def _assert(bool condition, str message):
>     assert condition, message
> ```
> 
> (also add the tensor overload)

Sorry, I'm not familiar at all with JIT frontend.  Seeing the following error (https://gist.github.com/vkuzo/75413cdc46c87880a39c4056228ebeda) when trying out your suggestion.  Thoughts?

eellison commented:
That’s not a JIT issue, you just have a circular import. Try defining _assert in _jit_internals

vkuzo committed:
comments on "torch._assert: make it torch.jit.script'able"


Summary:

Currently `torch._assert` is not scriptable, which makes it not very useful for production code.  According to @jamesr66a , moving this to c++ op land will help with scriptability.  This PR implements the change.

Note: with the current code the `_assert` is scriptable but is a no-op after being scripted.  Would love suggestions on how to address that (can be in future PR).

Test Plan:

```
python test/test_utils.py TestAssert.test_assert_scriptable
python test/test_utils.py TestAssert.test_assert_true
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

vkuzo committed:
rebase on "torch._assert: make it torch.jit.script'able"


Summary:

Currently `torch._assert` is not scriptable, which makes it not very useful for production code.  According to @jamesr66a , moving this to c++ op land will help with scriptability.  This PR implements the change.

Note: with the current code the `_assert` is scriptable but is a no-op after being scripted.  Would love suggestions on how to address that (can be in future PR).

Test Plan:

```
python test/test_utils.py TestAssert.test_assert_scriptable
python test/test_utils.py TestAssert.test_assert_true
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

eellison reviewed (approved):
lgtm, thanks for making the changes!

vkuzo committed:
rebase on "torch._assert: make it torch.jit.script'able"


Summary:

Currently `torch._assert` is not scriptable, which makes it not very useful for production code.  Adding a JIT override to address this.

Test Plan:

```
python test/test_utils.py TestAssert.test_assert_scriptable
python test/test_utils.py TestAssert.test_assert_true
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

vkuzo committed:
lint on "torch._assert: make it torch.jit.script'able"


Summary:

Currently `torch._assert` is not scriptable, which makes it not very useful for production code.  Adding a JIT override to address this.

Test Plan:

```
python test/test_utils.py TestAssert.test_assert_scriptable
python test/test_utils.py TestAssert.test_assert_true
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

vkuzo committed:
rebase on "torch._assert: make it torch.jit.script'able"


Summary:

Currently `torch._assert` is not scriptable, which makes it not very useful for production code.  Adding a JIT override to address this.

Test Plan:

```
python test/test_utils.py TestAssert.test_assert_scriptable
python test/test_utils.py TestAssert.test_assert_true
python test/test_fx.py TestFX.test_symbolic_trace_assert
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24740727](https://our.internmc.facebook.com/intern/diff/D24740727)

[ghstack-poisoned]

zou3519 commented:
This broke the docs build on master (and the docs build was failing on the PR, see https://app.circleci.com/pipelines/github/pytorch/pytorch/238714/workflows/ccc182c1-5983-4054-baac-b29f741cb111/jobs/8903580) so I am reverting this
